File: Part_1/Arrays/stacks/leetcode/valid_prentasis.cpp
#include <iostream>
#include <string>
#include <stack>
#include <unordered_map>

bool isValid(std::string &stringOfPrentasis)
{
  std::stack<char> stack;
  std::unordered_map<char, char> closeToOpen = {
      {'}', '{'},
      {']', '['},
      {')', '('}};

  for (char c : stringOfPrentasis)
  {
    if (closeToOpen.count(c))
    {
      if (!stack.empty() && stack.top() == closeToOpen[c])
      {
        stack.pop();
      }
      else
      {
        return false;
      }
    }
    else
    {
      stack.push(c);
    }
  }

  return stack.empty();
}

bool isValid_2(std::string &s)
{
  while (true) {
    std::size_t pos = std::string::npos;

    if ((pos = s.find("()")) != std::string::npos) {
      s.erase(pos, 2);
      continue;
    }

    if ((pos = s.find("[]")) != std::string::npos) {
      s.erase(pos, 2);
      continue;
    }

    if ((pos = s.find("{}")) != std::string::npos) {
      s.erase(pos, 2);
      continue;
    }

    break;
  }

  return s.empty();
}

int main()
{
  // std::string brackets = "]";
  std::string brackets = "([{}])";
  // std::string brackets = "()[]{}";
  // std::string brackets = "([)]]{}";
  // std::string brackets = "(){}}{";

  std::cout << std::boolalpha << isValid_2(brackets) << std::endl;

  return 0;
}

File: Part_1/Arrays/stacks/leetcode/valid_prentasis.py
def isValid_Stack(str: str) -> bool:
  stack = [];
  closeToOpen = { ')': '(', ']': '[', '}': '{'}
  
  for c in str:
    if c in closeToOpen:
      if stack.count != 0 and stack[-1] == closeToOpen[c]:
        stack.pop()
      else:
        return False
    else:
      stack.append(c);
  
  return len(stack) == 0;

def isValid_BruteForse(str: str) -> bool:
  while True:
    pos = -1
    
    if "()" in str:
      str = str.replace("()", "", 1)
      continue;
    
    if "[]" in str:
      str = str.replace("[]", "", 1)
      continue
    
    if "{}" in str:
      str = str.replace("{}", "", 1)
      continue
    
    break
  
  return len(str) == 0; 

str_test = "[{()}]"

print(isValid_BruteForse(str_test))